File: leetcode/2905.找出满足差值条件的下标 II.py
import sortedcontainers
from sortedcontainers import *
class Solution:
    def findIndices(self, nums: List[int], idf: int, vdf: int) -> List[int]:
        n = len(nums)
        st = SortedList()
        for i in range(n - idf):
            j = i + idf
            st.add((nums[i], i))
            if abs(st[0][0] - nums[j]) >= vdf:
                return [st[0][1], j]
            if abs(st[-1][0] - nums[j]) >= vdf:
                return [st[-1][1], j]
        return [-1, -1]

File: leetcode/2905.找出满足差值条件的下标 II.cpp
class Solution {
public:
    vector<int> findIndices(vector<int>& nums, int idf, int vdf) {
        int n = nums.size();
        std::set<std::pair<int, int>> st;

        for (int i = 0; i < n - idf; ++i) {
            int j = i + idf;
            st.insert(std::make_pair(nums[i], i));

            if (std::abs(st.begin()->first - nums[j]) >= vdf) {
                return {st.begin()->second, j};
            }

            if (std::abs(std::prev(st.end())->first - nums[j]) >= vdf) {
                return {std::prev(st.end())->second, j};
            }
        }

        return {-1, -1};
    }
};